Support paste verification

puntope opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

Closes #1001 .

Added functionality in order to allow the user to paste the code inside the verification code control inputs.

### Screenshots:

<!--- Optional --->

https://user-images.githubusercontent.com/5908855/143000769-1969f4b8-b0ba-4a24-9f63-f14544e8c1dd.mov



### Changes

* Mainly a refactor in verification Code Control
* Added a new `handlePaste` function to handle the input in the case is a paste event
* Added some tests  

### Detailed test instructions:

1. Go to Onboarding Setup Dashboard or to Settings -> Edit Phone
2. Edit the phone and click on Verify Phone Number
3. Check you can paste digits on the inputs 
4. Check you can insert the input as normal
5. You cannot insert letters or symbols, only digits. 

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

> Improve accessibility in Verification Code Control

puntope committed:
Adding Paste functionality tests for Verification Code Control

puntope committed:
Adding Paste functionality for Verification Code Control

puntope committed:
Update typo in comment

puntope committed:
Remove unnecessary ref

puntope committed:
Add tests for onCodeChange callback in Verification Code Control

eason9487 commented on js/src/components/contact-information/phone-number-card/verification-code-control.js:
```diff
@@ -135,13 +174,17 @@ export default function VerificationCodeControl( {
+			updateInputRefs( digits );
 		}
-	}, [ resetNeedle, digits ] );
+	}, [ digits ] );
```
❓ 
This `useEffect` was used to deal with the special workaround as the above code comments. I'd not suggest adding the handling of updating `digits` here. In addition, after changing the logic, the original automatic focus for input cursor is not working.

### Original
When no verification code is entered, the first input will get focused after
1. clicking the resend
2. switching the verification methods

![Kapture 2021-11-24 at 18 56 56](https://user-images.githubusercontent.com/17420811/143226739-cd0f1437-8184-4d7c-904b-17200a0f6a1e.gif)

### Current

Both cases are not working now
![Kapture 2021-11-24 at 18 55 37](https://user-images.githubusercontent.com/17420811/143227687-0a7e6a1d-cd25-4f7e-abcb-39c14710319c.gif)

eason9487 commented on js/src/components/contact-information/phone-number-card/verification-code-control.js:
```diff
@@ -86,36 +104,57 @@ export default function VerificationCodeControl( {
-		setDigits( initDigits );
-		onCodeChangeRef.current( toCallbackData( initDigits ) );
+		updateInputRefs( initDigits );
+		updateState( initDigits );
```
❓ 
Even though this `useEffect` result is correct, the `react-hooks/exhaustive-deps` warning messages would cause confusion when reading codes, and might further repeatedly check if there are any edge cases that could cause problems.

![image](https://user-images.githubusercontent.com/17420811/143228014-26291f28-2baf-413c-be14-ebcf297f8e7a.png)

eason9487 commented on js/src/components/contact-information/phone-number-card/verification-code-control.js:
```diff
@@ -86,36 +104,57 @@ export default function VerificationCodeControl( {
+			: handleInput( e );
+
+		maybeMoveFocus( nextFocusIdx );
+		updateState( nextDigits );
```
❓ 
The original implementation only called `onCodeChange` when digits did change, or reset by `resetNeedle`, but now it's called even no change has occurred. The timing of the callback trigger deviates from its meaning and description.

eason9487 commented on js/src/components/contact-information/phone-number-card/verification-code-control.js:
```diff
@@ -86,36 +104,57 @@ export default function VerificationCodeControl( {
+		// only allow n digits, from the current idx position until the end
+		const newDigits = [
+			...value.replace( /\D/g, '' ).substr( 0, DIGIT_LENGTH - idx ),
+		];
```
❓ 
We don't need to handle content other than plain text. Not sure why it needs to use `onPaste` event? The same `newDigits` value could be got in the original `onInput` handler by 
```js
const handleInput = ( e ) => {
	const { value, dataset } = e.target;
	const newDigits = value
		.replace( /\D/g, '' )
		.substr( cursorRef.current, DIGIT_LENGTH - idx );

	// ...omitted
};
```

I think this could simplify the processing of getting content from `onPaste`, and even further simplify other changes by using the original `handleInput` to handle a 1 to `DIGIT_LENGTH` length array of digits.

eason9487 reviewed (commented):
Sorry I haven't looked at the **.test.js** part. This review only comments on the implementation first.

1. 🚧   When pasting, the `<input>` value of the last digit is incorrectly displayed back to the previous value till losing focus.
    ![Kapture 2021-11-24 at 17 33 04](https://user-images.githubusercontent.com/17420811/143242578-c6aed09c-2f5a-4a91-b9c7-fec0bc333988.gif)
    ![Kapture 2021-11-24 at 17 35 07](https://user-images.githubusercontent.com/17420811/143242681-e1caa9f7-282a-43e2-8c95-522e3529bfec.gif)
2. ❓  When the length of the pasted codes are not long enough to fill to the last `<input>`, it should be better to move the focus to the next `<input>` to be entered
    ![Kapture 2021-11-24 at 17 42 31](https://user-images.githubusercontent.com/17420811/143242725-23fc5dc4-0698-4817-9eb9-8c70c6e12794.gif)

eason9487 commented:
FYI - If need to get more context from the previous relevant discussions, you can find them in these PRs:
- #993
- #1007

puntope committed:
Fixes unnecessary state updates, deps warning and the focus issue on reset needle.

puntope committed:
Improve focus logic on paste

puntope commented:
> 2\. ❓  When the length of the pasted codes are not long enough to fill to the last `<input>`, it should be better to move the focus to the next `<input>` to be entered

Agree! 👍 

Fixed in [e37f3b7](https://github.com/woocommerce/google-listings-and-ads/pull/1106/commits/e37f3b712305d29ff9d6d294fd4bff114ca86920)

puntope commented:
> 1. 🚧   When pasting, the `<input>` value of the last digit is incorrectly displayed back to the previous value till losing focus.

This is definitely working for me. 😕   Could you double-check? 

Note: At some point in the dev process, it wasn't working, that is why I added the statement `updateInputRefs( digits );`

eason9487 commented:
> > 1. 🚧   When pasting, the `<input>` value of the last digit is incorrectly displayed back to the previous value till losing focus.
> 
> This is definitely working for me. 😕 Could you double-check?

Tried with `npm install`, restarting `npm start`, and the latest version of Chrome, the problem can be reproduced on my end. A new finding is that it only occurs during the cooling countdown, which means that it probably occurs when any re-renders are triggered from upper layers.

![Kapture 2021-11-25 at 11 05 10](https://user-images.githubusercontent.com/17420811/143372811-5098db39-02f4-41da-9b6d-31c9fc2f7887.gif)

puntope committed:
Fix focus sync issue in Validaton Control

puntope committed:
Fix reset needle conflicts

puntope commented:
> A new finding is that it only occurs during the cooling countdown, which means that it probably occurs when any re-renders are triggered from upper layers.

First of all, thanks for catching this @eason9487! 🙏  My manual test was wrong cos my counter was off due request limit. Now I was able to reproduce and fix the error.

> the problem

**Bug causes:**

Testing deeper I can confirm the problem is related to the way `input-field` handles the internal state on focus after each render. The value is not updated in the render until the input lost focus. I could confirm that everything works fine replacing that ìnput-field` with a native input.

On single-digit insertion, we don't have that problem since we move the focus to the next input box. Hence, the previous one is updated correctly after losing focus.


**Solution:**

I replaced now the `inputRefs` update solution with a new state solution, in order to create an extra tick and sync the internal state for the `input-field`  

Besides that, I refactored a bit the `resetNeedle` effects in order to avoid sync conflicts.

Let me know if it works now :)  


**Tested:**

* Works with counter running both insert and paste, full and partial pastes.
* Change on resetNeedle reset the digits and the focus. (Click on resend, change method, request error)
* Works with counter stoped both insert and paste, full and partial pastes.
* Verification code value in parent component is updated.
* No es-lint warns
* Unit tests passed

eason9487 commented:
Thanks for the updates!

However, I'm sorry that it brings new issues and some previous issues are still unresolved.

1. 🚧  The original automatic focus for input cursor is still not working when no entered verification codes.
    **Switch verification method**
    ![Kapture 2021-11-25 at 18 12 40](https://user-images.githubusercontent.com/17420811/143424620-00ce60a0-c015-49de-b50d-3c62d2c64c8b.gif)
    **Resend**
    ![Kapture 2021-11-25 at 18 11 54](https://user-images.githubusercontent.com/17420811/143424538-0d05377d-e133-40dc-a86e-32ed591fce9b.gif)
2. 🚧  Sometimes the focus stays at the pasting `<input>` with previous value when pasting filled `<input>`s. Steps I tried
    1. Paste `123456` on the first `<input>`
    2. Copy `886` and pate on the third `<input>`
    3. The displaying values are changed to `123866`, but I expected it would be `128866` and the focus should be moved to the last `<input>`.
    ![Kapture 2021-11-25 at 18 17 17](https://user-images.githubusercontent.com/17420811/143425447-182e8117-a991-47de-8029-fb28d994e0bc.gif)
3. 🚧  When pasting on the last <input>, the problem of the display value is out of sync with the internal state still exists till losing focus.
    ![Kapture 2021-11-25 at 18 23 00](https://user-images.githubusercontent.com/17420811/143426029-8ec62142-751e-426f-8299-c0436413fd47.gif)

eason9487 commented on js/src/components/contact-information/phone-number-card/verification-code-control.js:
```diff
@@ -81,42 +89,70 @@ export default function VerificationCodeControl( {
+
+		setFocus( nextFocusIdx );
+
+		if ( nextDigits !== digits ) {
+			updateState( nextDigits );
+		}
```
🚧 
Use `===` to check whether the values in two different arrays are equal is not appropriate.

eason9487 commented on js/src/components/contact-information/phone-number-card/verification-code-control.js:
```diff
@@ -81,42 +89,70 @@ export default function VerificationCodeControl( {
 		}
 	};
 
+	const updateState = useCallback(
+		( nextDigits ) => {
+			setDigits( nextDigits );
+			onCodeChange( toCallbackData( nextDigits ) );
+		},
+		[ onCodeChange ]
+	);
```
🚧 
It would cause infinite re-rendering and locking up the entire web page if the consumer side passes `onCodeChange` with a callback that changes its reference every time and will also trigger re-render on the topper layer. For example, replace the `onCodeChange={ setVerification }` with `onCodeChange={ ( code ) => setVerification( code ) }` at:

https://github.com/woocommerce/google-listings-and-ads/blob/0d0f77111815a85e7303785c64d4808d9760f951/js/src/components/contact-information/phone-number-card/verify-phone-number-content.js#L193-L196


The `useRef` removed by https://github.com/woocommerce/google-listings-and-ads/commit/875f05ecc538fa69694beaeda96af880c4c7018c is not unnecessary. It was used for this reason.

eason9487 reviewed (changes requested):
There're two new critical issues after adding new commits.

puntope committed:
Fixing reset needle focus issues and edge cases in Validation Code Control

puntope commented:
> However, I'm sorry that it brings new issues and some previous issues are still unresolved.

No sorry, thanks to you for checking this! Better you than the users :) Thank you for being so thorough in the reviews!

> 🚧 The original automatic focus for input cursor is still not working when no entered verification codes.
Switch verification method

Fixed here [d20faae](https://github.com/woocommerce/google-listings-and-ads/pull/1106/commits/d20faae35834fa79a924f6b4254773c25c4dbf75) 


> 🚧 Sometimes the focus stays at the pasting <input> with previous value when pasting filled <input>s. Steps I tried
Paste 123456 on the first <input>
Copy 886 and pate on the third <input>
The displaying values are changed to 123866, but I expected it would be 128866 and the focus should be moved to the last <input>.

 Fixed here [d20faae](https://github.com/woocommerce/google-listings-and-ads/pull/1106/commits/d20faae35834fa79a924f6b4254773c25c4dbf75) 
 
 
> 🚧 When pasting on the last , the problem of the display value is out of sync with the internal state still exists till losing focus.

 Fixed here [d20faae](https://github.com/woocommerce/google-listings-and-ads/pull/1106/commits/d20faae35834fa79a924f6b4254773c25c4dbf75)

puntope commented:
> The useRef removed by 875f05e is not unnecessary. It was used for this reason.

Restored!

puntope commented:
💅 

> Use === to check whether the values in two different arrays are equal is not appropriate.

Added toString() strategy. Other alternatives are...

_.isEqual (lodash)
Just a loop
Stringify

eason9487 commented on js/src/components/contact-information/phone-number-card/verification-code-control.js:
```diff
@@ -81,42 +96,77 @@ export default function VerificationCodeControl( {
+			? handlePaste( e )
+			: handleInput( e );
+
+		setFocus( nextFocusIdx );
```
🚧 

The `focus` state may be out of sync with the actual focusing index of `<input>` nodes, and further causes the input to get stuck due to all `deps` have not been changed here:

https://github.com/woocommerce/google-listings-and-ads/blob/d20faae35834fa79a924f6b4254773c25c4dbf75/js/src/components/contact-information/phone-number-card/verification-code-control.js#L198-L200

Steps:

1. Enter a number in the first `<input>`
2. Move back to the first `<input>` by the ← key or mouse
3. Enter the same number as step 1
4. I expect it would move to the next input but it gets stuck

![Kapture 2021-11-29 at 17 57 06](https://user-images.githubusercontent.com/17420811/143846983-cd5fcc53-04c4-44a5-ac3a-02d96020ff4c.gif)

eason9487 commented on js/src/components/contact-information/phone-number-card/verification-code-control.js:
```diff
@@ -81,42 +96,77 @@ export default function VerificationCodeControl( {
-		// If that char is not a digit, then clear the input to empty.
 		const digit = value.substr( cursorRef.current, 1 ).replace( /\D/, '' );
+
+		// If that char is not a digit, then clear the input to empty.
```
❓ 
The meaning of this code comment is different from the `if ( digit !== value )` condition and its processing below.

eason9487 commented on js/src/components/contact-information/phone-number-card/verification-code-control.js:
```diff
@@ -135,13 +185,19 @@ export default function VerificationCodeControl( {
 	 * So here we await the `digits` is reset back to `initDigits` by above useEffect and sync to internal value,
```
💅 
The code comments are outdated after changing the position and implementation of the original `useEffect`.

eason9487 commented on js/src/components/contact-information/phone-number-card/verification-code-control.js:
```diff
@@ -81,42 +96,77 @@ export default function VerificationCodeControl( {
+			...value.replace( /\D/g, '' ).substr( 0, DIGIT_LENGTH - idx ),
+		];
+
+		// edge case: blur when pasting on last item
+		if (
+			newDigits.length === 1 &&
+			newDigits[ 0 ] !== digits[ idx ] &&
+			idx === DIGIT_LENGTH - 1
+		) {
+			e.target.blur();
```
📜 
I think we don't really need this. I will propose an alternative that simplifies overall implementations by the new submitting code review.

eason9487 reviewed (commented):
📜 
Due to the internal state management of `<InputControl>`, the implementation of this component has to use some workaround for the focus controlling of input cursor. But the fixing result after several rounds of code review makes the implementation start to become too complex. Even more, workarounds have to be used at more places because of other side effects caused by the added workarounds.

The key problem is when a `<InputControl>` has the focus, updating the `value` prop would not affect the actually displaying content on the `<input>`, and even update via the value property of the real `<input>` DOM node could temporarily show the wanted content, it will be still recovered back to the content of passing `value` prop after any re-render occurs.

When we want to set the focus on an unfocused `<InputControl>`, which is also updating its display content, this state conflict happened. To deal with this, it could be resolved by updating the `value` prop and yielding the focus move to the next tick of re-render.

And this case will only happen in the first (when reset) and the last (when pasted) `<InputControl>`. Moving the workaround into the `maybeMoveFocus` function with a bit of magic could make the implementation simpler and clearer.

```js
const maybeMoveFocus = ( targetIdx ) => {
	const validIdx = Math.max( Math.min( targetIdx, DIGIT_LENGTH - 1 ), 0 );
	const node = inputsRef.current[ validIdx ];

	if ( targetIdx === validIdx ) {
		node.focus();
	} else {
		setTimeout( () => node.focus() );
	}
};
```

Then we could call `maybeMoveFocus( -1 );` when resetting, and `maybeMoveFocus( DIGIT_LENGTH );` when moving focus to the last. With this method, it will need to change most other implementations. I made the alternative with code comments here https://github.com/woocommerce/google-listings-and-ads/compare/8db08347d20b0f16178edba75899c2283c5877db...707e15bb3826e90e929ebb824a3b78e4d3af5198. Probably we could consider this solution.

eason9487 commented:
Closing this PR as the phone verification has been removed by #2653.
